src/cockpit/packages.py

```diff
@@ -15,193 +15,247 @@
-        translation = translations.get(locale)
-        if translation is not None:
-            return translation
+        manifest_policy = self.manifest.get('content-security-policy', '')
+        if not isinstance(manifest_policy, str):
+            manifest_policy = ''
+        for item in manifest_policy.split(';'):
+            item = item.strip()
+            if item:
+                key, _, value = item.strip().partition(' ')
+                policy[key] = value
 
-    # Next, check the language-only versions of variant-specified locales
-    for locale in locales:
-        language, _, region = locale.partition('-')
-        if not region:
-            continue
-        translation = translations.get(language)
-        if translation:
-            return translation
+        return ' '.join(f'{k} {v};' for k, v in policy.items()) + ' block-all-mixed-content'
 
-    # If nothing else worked, we always have English
-    return translations['en']
+    def load_file(self, filename: str) -> Document:
+        path = self.path / filename
+        logger.debug('  loading data from %s', path)
+        data = path.read_bytes()
 
+        content_type, content_encoding = mimetypes.guess_type(filename)
+        content_security_policy = None
 
-class PackagesListener:
-    def packages_loaded(self):
-        """Called when the packages have been reloaded"""
+        if content_type is None:
+            content_type = 'text/plain'
+        elif content_type.startswith('text/html'):
+            content_security_policy = self.get_content_security_policy()
 
+        return Document(data, content_type, content_encoding, content_security_policy)
 
-class Package:
-    CONDITIONS: ClassVar[Dict[str, Callable[[str], bool]]] = {
-        'path-exists': os.path.exists,
-        'path-not-exists': lambda p: not os.path.exists(p),
-    }
+    def load_translation(self, locales: List[str]) -> Document:
+        self.ensure_scanned()
+        assert self.translations is not None
 
-    # For po.js files, the interesting part is the locale name
-    PO_JS_RE: ClassVar[Pattern] = re.compile(r'po\.([^.]+)\.js(\.gz)?')
+        # First check the locales that the user sent
+        for locale in locales:
+            with contextlib.suppress(KeyError):
+                return self.load_file(self.translations[locale])
 
-    # A built in base set of "English" translations
-    PO_EN_JS: ClassVar[Entity] = (b'', mimetypes.guess_type('po.js'))
-    BASE_TRANSLATIONS: ClassVar[Entities] = {'en': PO_EN_JS, 'en-us': PO_EN_JS}
+        # Next, check the language-only versions of variant-specified locales
+        for locale in locales:
+            language, _, region = locale.partition('-')
+            if region:
+                with contextlib.suppress(KeyError):
+                    return self.load_file(self.translations[language])
 
-    files: Entities
-    translations: Entities
+        # We prefer to return an empty document than 404 in order to avoid
```

Comment:
:+1: these were annoying indeed

Reply:
I really like the new `Document` type here...